### PR TITLE
WIP: Citation metadata w/ duecredit

### PR DESCRIPTION
**Pull request by cgranade**
This WIP PR addresses #128 by adding citation metadata to ``__init__.py``. This citation is guarded with the tag ``["implementation"]``, with the idea that citations to underlying theory could be added to individual functions with the ``["theory"]`` tag.

### CODE REVIEW

**Comment by ihincks** [comment]
Looking good, Chris. Are you thinking of implementations=experimental, or having both? It would also be good to add something to the guide about how to use it.

**Comment by cgranade** [comment]
Thanks! My thought was to have tags broken down into ``theory`` and ``implementations``, with the latter being included by duecredit by default, but I really like your suggestion of ``experiment`` as well to point to experimental applications. I'll start adding to the guide on how to use the new duecredit functionality soon, with the explicit note that the bibliography is not yet complete and that users should still be careful to cite as appropriate.

**Comment by cgranade** [comment]
One weird bug so far: ``duecredit`` injects "Frey, B.J. & Dueck, D., 2007. Clustering by Passing Messages Between Data Points. Science, 315(5814), pp.972–976." as a citation for a part of scikit-learn's clustering library, such that that citation shows up if the user has scikit-learn installed. I'm not sure quite what to do about that, since QInfer definitely does ``import sklearn.cluster`` in ``clustering.py``.

**Comment by ihincks** [comment]
Let me know when you satisfied with the state of this PR. I will test it on my machine, and subsequently provide a code review.

**Comment by cgranade** [comment]
I think I'm mostly happy with this PR now, modulo that the process listed in the docs for generating a bibliography using the ``DUECREDIT_ENABLE="yes"`` environment variable setting seems to not always work. I'm a bit confused as to exactly what's going on there, and it could well be upstream, but if you have any insight I'd greatly appreciate it. Thanks! ♥

**Comment by ihincks** [comment]
Played around briefly, seems to be working. If I just `import qinfer as qi` without anything else (whether using env variable or with `python -m duecredit`), I end up with 

    [4] 2017-04-03 11:16:13,396 [WARNING] DueCredit internal failure while running <function _dump_collector_summary at 0x7fc3c2755aa0>: OSError(13, 'Permission denied'). Please report to developers at https://github.com/duecredit/duecredit/issues (utils.py:199)

Do you get this too?

I am also wondering about jupyter notebooks. It seems that you can get the current state of collection using:

    import os
    os.environ['DUECREDIT_ENABLE'] = 'yes'
    import qinfer as qi
    qi.due._dump_collector_summary()

Although from my usual use of qinfer (through notebooks, possibly with local modules) this seems convenient, I would hesitate to recommend people use hidden functions.

**Comment by ihincks** [comment]
Submitted issue: https://github.com/duecredit/duecredit/issues/104
and: https://github.com/duecredit/duecredit/issues/105

**Comment by cgranade** [comment]
Thanks for looking at that! I didn't get that error, but rather intermittently got a ``UnicodeDecodeError``. I'll isolate a test case and file it as an issue with DueCredit as well.

File: doc/source/guide/citing.rst [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    [2] Granade, C. et al., 2016. QInfer: Statistical Inference Software for Quantum Applications. arXiv:1610.00336 [physics, physics:quant-ph, stat].
+    [3] Pedregosa, F. et al., 2011. Scikit-learn: Machine learning in Python. The Journal of Machine Learning Research, 12, pp.2825–2830.
+    [4] Frey, B.J. & Dueck, D., 2007. Clustering by Passing Messages Between Data Points. Science, 315(5814), pp.972–976.
+
+The bibliography entries defined by QInfer are organized according to different
+*tags*, making it easier to filter through the results of `duecredit`_. In particular,
+QInfer defines the following citation tags:
+
+- ``implementation``: Concerns a software implementation of an algorithm or computation.
+- ``theory``: Concerns the initial proposal of an algorithm and other theory results.
```
Comment (yarikoptic) [inline doc/source/guide/citing.rst]: interesting... so how this is different from implementation really?  is it when you are to provide a "reference-implementation" of a new method (may be not yet published, thus may be no reference even yet) which rests on this "theory" publication, so not implementing but rather extending etc upon it?

meanwhile I've decided to accumulate those tags within our README.md: https://github.com/duecredit/duecredit/pull/109 . Feel welcome to chime in ;)

Comment (cgranade) [inline doc/source/guide/citing.rst]: Thanks for the feedback! I'll cross-post to duecredit/duecredit#109, but the rough version is that some of what we tagged as ``"theory"`` doesn't come with an implementation at all. For instance, the [Bruzda *et al.*](https://dx.doi.org/10.1016/j.physleta.2008.11.043) proposes a novel algorithm, but doesn't provide source code. In providing an implementation, we don't want to claim credit for the underlying theoretical proposal of the Bruzda *et al.* algorithm, and have used a ``"theory"`` tag to indicate that.

Comment (yarikoptic) [inline doc/source/guide/citing.rst]: well, it is unlikely that there is/was any algorithm without any (in code, or hardware of some kind) implementation ;)    But we do have `reference implementation` to describe that very original one, and all the others are just `implementation` of the cited algorithm.
So the difference here is also somewhat of directionality.  You use tag "theory" to point that "that paper provides the theoretical explanation for this implementation".  We say "implementation" as "this is an implementation of the theory in that paper".  So kinda the same in principle.

**Comment by ihincks** [comment]
Hmm, the test script

```
import qinfer as qi
m = qi.CoinModel()
prior = qi.UniformDistribution([[0,1]])
u = qi.SMCUpdater(m, 1000, prior)
u.resampler(m, u.particle_weights, u.particle_locations)
```

is not putting Lui and West in the list. Am I doing something wrong? Just running `python -m duecredit test.py`.

PS: the author tag looks a bit funny: `author = {Liu, Jane and West, Mike},`

**Comment by cgranade** [comment]
The ``@dcite`` annotation for the Liu–West resampler uses the ``["theory"]`` tag, which is disabled by default. Does the test script work with ``DUECREDIT_REPORT_TAGS=*``?

File: doc/source/guide/citing.rst (hunk 1 above)
Comment (cgranade) [inline doc/source/guide/citing.rst]: Ah, I think I understand better, then. Am I right in reading ``"implementation"`` as roughly saying "implementation of"? In which case, I agree, our use of ``"theory"`` in this PR coincides with the designed use of ``"implementation"``. I'm happy to change our PR, then, to match the documentation in duecredit/duecredit#109.

Comment (yarikoptic) [inline doc/source/guide/citing.rst]: yes, "implementation of" ;-)
the advantage of using this one is that the default tags used (unless 'DUECREDIT_REPORT_TAGS' is specified) are "implementation" and "reference-implementation" so duecredit would provide references for those.  This way if someone uses a method implemented in a given software, users would get citation to use for that method since they have used it.
hopefully this all makes sense ;)
